jq255.md

```diff
@@ -0,0 +1,1425 @@
+    e1e2 = E1*E2
+    z1z2 = Z1*Z2
+    u1u2 = U1*U2
+    t1t2 = T1*T2
+    tz = (Z1 + T1)*(Z2 + T2) - z1z2 - t1t2
+    eu = (E1 + U1)*(E2 + U2) - e1e2 - u1u2
+    hd = z1z2 - bp*t1t2
+    E3 = (z1z2 + bp*t1t2)*(e1e2 + ap*u1u2) + 2*bp*u1u2*tz
+    Z3 = hd**2
+    T3 = eu**2
+    U3 = half((hd + eu)**2 - Z3 - T3)   # Alternatively: U3 = hd*eu
+    return (E3, Z3, U3, T3)
+```
+
+If the point `P2` is in extended affine coordinates, i.e. such that `Z2`
+is statically known to be equal to 1, then the multiplication `Z1*Z2`
+can be avoided, and the computation of `tz` simplifies to `Z1*T2 + T1`.
+This specific case is known as a *mixed addition* and is typically
+encountered when multiplying the conventional generator by a scalar,
+using precomputed tables of small multiples of the generator.
+
+### Point Negation and Subtraction
+
+Given the point `P = (E:Z:U:T)`, its opposite is obtained as
+`-P = (E:Z:-U:T)`, i.e. simply negating the `U` coordinate in the field.
+The subtraction of `P2` from `P1` can be obtained by adding `-P2` to
+`P1`.
+
+```python
+def point_neg(P):
+    (E, Z, U, T) = P
+    return (E, Z, -U, T)
+
+def point_sub(P1, P2):
+    return point_add(P1, point_neg(P2))
+```
+
+### Point Doublings
+
+Suppose that we have `k >= 1` successive doublings to perform on an
+input point `P` (i.e. we want `(2^k)*P`). The general strategy is
+the following:
+
+  - Perform the first doubling with formulas that receive `P` in
+    EZUT coordinates, and output `2*P` in XWJ coordinates.
+  - Perform the other `k-1` doublings in XWJ coordinates.
+  - Convert the final result from XWJ to EZUT.
+
+Depending on the used curve, it can be beneficial to combine some of
+these doublings with a switch to the other representant (i.e. computing
+`2*P+N` instead of `2*P`); this does not change the output as a group
+element, since we are free to use either of the two representants.
+
+Two distinct functions, described below, are used for computing sequences
+of doublings on jq255e and jq255s, each leveraging the specific definition
+of the curve to achieve better performance. The subsequent pseudocode
+will assume that the correct implementation for the current curve is
+accessible under the generic function name `point_xdouble()`.
+
+#### On jq255e
+
+Given a jq255e element represented by point `P`, and integer `k >= 1`,
+the following function computes `k` successive doublings:
+
+```python
+def point_xdouble_jq255e(P, k):
+    (E, Z, U, T) = P
+
+    # P (EZUT) -> 2*P (XWJ)
+    s = E**2
+    X = s**2
+    W = 2*(Z**2) - s
+    J = 2*(E*U)
+
+    # k-1 times P (XWJ) -> 2*P (XWJ)
+    for _ in range(1, k):
+        s1 = W**2
+        s2 = s1 - 2*X
+        s3 = s2**2
+        X = s3**2
+        J = J*((W + s2)**2 - s1 - s3)   # Alternatively: J = 2*J*W*s2
+        W = s3 - 2*(s1**2)
+
+    # Conversion XWJ -> EZUT
+    Z = W**2
+    T = J**2
+    U = half((W + J)**2 - Z - T)        # Alternatively: U = J*W
+    E = 2*X - Z
```

Comment:
Should this be its own subroutine?

Reply:
I kept it inlined for two (weak) reasons:
1. It avoids having (X,W,J) triplets moving around in the API. It can lead to confusion if there are several competing systems of coordinates, and I prefer to keep them contained inside the xdouble function.
2. In practice (and in more aggressively optimized languages), keeping them inline helps substantially with performance.

All of this is debatable.

Reply:
Also: the XWJ -> EZUT conversions are distinct between jq255e and jq255s. The computation of `E` is really `E = 2*X - Z + a*T` and the two curves don't use the same `a` (`a = 0` for jq255e, and `a = 1` for jq255s), so that's an extra reason not to try to merge them into a single shared function.